src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    currentState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      Object.assign(currentState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const prop of action.keysToRemove) {
+        delete currentState[prop];
+      }
+    }
+
+    if (action.type === 'clear') {
+      currentState = {};
+    }
```

Comment:
Consider using a switch statement instead of multiple if statements for handling different action types. This will make the code cleaner and easier to maintain. Remember to include a default case for error handling as per the checklist.